### PR TITLE
guide: CI/CD with Tekton

### PR DESCRIPTION
**Pull request by jasonmadigan**
Re: https://github.com/Kuadrant/docs.kuadrant.io/issues/64

Steps to Run:

- Probably easiest to install Kuadrant via api-quickstart
- Switch to the workload1 cluster (which has the `kuadrant-operator` running)
- Run the steps

Sample output:

```bash
tkn taskrun logs -f run-kuadrantctl-taskrun -n petstore                               git:(64-kuadrantctl-tekton|…2⚑2

[clone] Cloning into '/workspace/source'...
[clone] Already on 'main'
[clone] Your branch is up to date with 'origin/main'.

[download-kuadrantctl]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[download-kuadrantctl]                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 21.4M  100 21.4M    0     0  12.7M      0  0:00:01  0:00:01 --:--:-- 20.9M

[run-kuadrantctl] kind: AuthPolicy
[run-kuadrantctl] apiVersion: kuadrant.io/v1beta2
[run-kuadrantctl] metadata:
[run-kuadrantctl]   name: petstore
[run-kuadrantctl]   namespace: petstore
[run-kuadrantctl]   creationTimestamp: null
[run-kuadrantctl]   labels:
[run-kuadrantctl]     deployment: petstore
[run-kuadrantctl]     owner: jbloggs
[run-kuadrantctl] spec:
[run-kuadrantctl]   targetRef:
[run-kuadrantctl]     group: gateway.networking.k8s.io
[run-kuadrantctl]     kind: HTTPRoute
[run-kuadrantctl]     name: petstore
[run-kuadrantctl]     namespace: petstore
[run-kuadrantctl]   routeSelectors:
[run-kuadrantctl]     - matches:
[run-kuadrantctl]         - path:
[run-kuadrantctl]             type: Exact
[run-kuadrantctl]             value: /api/v3/store/admin
[run-kuadrantctl]           method: GET
[run-kuadrantctl]   rules:
[run-kuadrantctl]     authentication:
[run-kuadrantctl]       storeAdmin_api_key:
[run-kuadrantctl]         credentials:
[run-kuadrantctl]           customHeader:
[run-kuadrantctl]             name: api_key
[run-kuadrantctl]         apiKey:
[run-kuadrantctl]           selector:
[run-kuadrantctl]             matchLabels:
[run-kuadrantctl]               kuadrant.io/apikeys-by: api_key
[run-kuadrantctl]         routeSelectors:
[run-kuadrantctl]           - matches:
[run-kuadrantctl]               - path:
[run-kuadrantctl]                   type: Exact
[run-kuadrantctl]                   value: /api/v3/store/admin
[run-kuadrantctl]                 method: GET
[run-kuadrantctl] status: {}
[run-kuadrantctl] kind: RateLimitPolicy
[run-kuadrantctl] apiVersion: kuadrant.io/v1beta2
[run-kuadrantctl] metadata:
[run-kuadrantctl]   name: petstore
[run-kuadrantctl]   namespace: petstore
[run-kuadrantctl]   creationTimestamp: null
[run-kuadrantctl]   labels:
[run-kuadrantctl]     deployment: petstore
[run-kuadrantctl]     owner: jbloggs
[run-kuadrantctl] spec:
[run-kuadrantctl]   targetRef:
[run-kuadrantctl]     group: gateway.networking.k8s.io
[run-kuadrantctl]     kind: HTTPRoute
[run-kuadrantctl]     name: petstore
[run-kuadrantctl]     namespace: petstore
[run-kuadrantctl]   limits:
[run-kuadrantctl]     getInventory:
[run-kuadrantctl]       routeSelectors:
[run-kuadrantctl]         - matches:
[run-kuadrantctl]             - path:
[run-kuadrantctl]                 type: Exact
[run-kuadrantctl]                 value: /api/v3/store/inventory
[run-kuadrantctl]               method: GET
[run-kuadrantctl]       rates:
[run-kuadrantctl]         - limit: 10
[run-kuadrantctl]           duration: 10
[run-kuadrantctl]           unit: second
[run-kuadrantctl]     loginUser:
[run-kuadrantctl]       routeSelectors:
[run-kuadrantctl]         - matches:
[run-kuadrantctl]             - path:
[run-kuadrantctl]                 type: Exact
[run-kuadrantctl]                 value: /api/v3/user/login
[run-kuadrantctl]               method: GET
[run-kuadrantctl]       rates:
[run-kuadrantctl]         - limit: 2
[run-kuadrantctl]           duration: 10
[run-kuadrantctl]           unit: second
[run-kuadrantctl] status: {}
[run-kuadrantctl] kind: HTTPRoute
[run-kuadrantctl] apiVersion: gateway.networking.k8s.io/v1beta1
[run-kuadrantctl] metadata:
[run-kuadrantctl]   name: petstore
[run-kuadrantctl]   namespace: petstore
[run-kuadrantctl]   creationTimestamp: null
[run-kuadrantctl]   labels:
[run-kuadrantctl]     deployment: petstore
[run-kuadrantctl]     owner: jbloggs
[run-kuadrantctl] spec:
[run-kuadrantctl]   parentRefs:
[run-kuadrantctl]     - kind: Gateway
[run-kuadrantctl]       namespace: kuadrant-multi-cluster-gateways
[run-kuadrantctl]       name: prod-web
[run-kuadrantctl]   hostnames:
[run-kuadrantctl]     - petstore.${KUADRANT_ZONE_ROOT_DOMAIN}
[run-kuadrantctl]   rules:
[run-kuadrantctl]     - matches:
[run-kuadrantctl]         - path:
[run-kuadrantctl]             type: Exact
[run-kuadrantctl]             value: /api/v3/user/login
[run-kuadrantctl]           method: GET
[run-kuadrantctl]       backendRefs:
[run-kuadrantctl]         - name: petstore
[run-kuadrantctl]           namespace: petstore
[run-kuadrantctl]           port: 8080
[run-kuadrantctl]     - matches:
[run-kuadrantctl]         - path:
[run-kuadrantctl]             type: Exact
[run-kuadrantctl]             value: /api/v3/store/admin
[run-kuadrantctl]           method: GET
[run-kuadrantctl]       backendRefs:
[run-kuadrantctl]         - name: petstore
[run-kuadrantctl]           namespace: petstore
[run-kuadrantctl]           port: 8080
[run-kuadrantctl]     - matches:
[run-kuadrantctl]         - path:
[run-kuadrantctl]             type: Exact
[run-kuadrantctl]             value: /api/v3/store/inventory
[run-kuadrantctl]           method: GET
[run-kuadrantctl]       backendRefs:
[run-kuadrantctl]         - name: petstore
[run-kuadrantctl]           namespace: petstore
[run-kuadrantctl]           port: 8080
[run-kuadrantctl] status:
[run-kuadrantctl]   parents: null

[apply-resources] authpolicy.kuadrant.io/petstore created
[apply-resources] httproute.gateway.networking.k8s.io/petstore created
[apply-resources] ratelimitpolicy.kuadrant.io/petstore created
```

### CODE REVIEW

**Comment by jasonmadigan** [comment]
@david-martin  draft. I need to do an e2e spin through

File: doc/kuadrantctl-ci-cd.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+      cd $(workspaces.source.path)
+      ./kuadrantctl generate kuadrant authpolicy --oas openapi.yaml | yq eval -P | tee generated-resources/authpolicy.yaml
+      ./kuadrantctl generate kuadrant ratelimitpolicy --oas openapi.yaml | yq eval -P | tee generated-resources/ratelimitpolicy.yaml
+      ./kuadrantctl generate gatewayapi httproute --oas openapi.yaml | yq eval -P | tee generated-resources/httproute.yaml
+   - name: apply-resources
+      image: lachlanevenson/k8s-kubectl
+      script: |
+        cd $(workspaces.source.path)
+        kubectl apply -f ./generated-resources -n kuadrantctl-ci-example
```
Comment (david-martin) [inline doc/kuadrantctl-ci-cd.md:96]:
```suggestion
  - name: apply-resources
    image: lachlanevenson/k8s-kubectl
    script: |
      cd $(workspaces.source.path)
      kubectl apply -f ./generated-resources -n kuadrantctl-ci-example
```
Indentation a little off here

**Review by david-martin — CHANGES_REQUESTED** [review]
Got the task to run, which some changes suggested above.
The task is failing at the moment. I haven't looked into why, but here's the status:

```
kubectl get taskrun -A
NAMESPACE                NAME                      SUCCEEDED   REASON                    STARTTIME   COMPLETIONTIME
kuadrantctl-ci-example   run-kuadrantctl-taskrun   False       TaskRunResolutionFailed   76s         76s
```

```
      message: 'error when listing tasks for taskRun run-kuadrantctl-taskrun: tasks.tekton.dev
        "run-kuadrantctl" not found'
```

events
```
kubectl -n kuadrantctl-ci-example get ev
LAST SEEN   TYPE      REASON                 OBJECT                                         MESSAGE
109s        Normal    Started                taskrun/run-kuadrantctl-taskrun
109s        Warning   Failed                 taskrun/run-kuadrantctl-taskrun                error when listing tasks for taskRun run-kuadrantctl-taskrun: tasks.tekton.dev "run-kuadrantctl" not found
109s        Warning   InternalError          taskrun/run-kuadrantctl-taskrun                1 error occurred:...
2m11s       Normal    WaitForFirstConsumer   persistentvolumeclaim/tekton-kuadrantctl-pvc   waiting for first consumer to be created before binding
```

**Comment by jasonmadigan** [comment]
@david-martin this is actually runnable now. Added some steps to test.

File: doc/kuadrantctl-ci-cd.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+Create an associated `ClusterRole` and `ClusterRoleBinding`:
+
+```bash
+kubectl apply -n petstore -f - <<EOF
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: kuadrant-ci-example-full-access
+rules:
+- apiGroups: ["*"]
```
Comment (jasonmadigan) [inline doc/kuadrantctl-ci-cd.md:132]: could probably narrow these

File: doc/kuadrantctl-ci-cd.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        x86_64) BIN_ARCH="amd64";;
+        arm64) BIN_ARCH="arm64";;
+        aarch64) BIN_ARCH="arm64";;
+        *) echo "Unsupported architecture: $ARCH" && exit 1 ;;
+        esac
+        cd $(workspaces.source.path)
+        curl -LO "https://github.com/Kuadrant/kuadrantctl/releases/download/v0.2.0/kuadrantctl-v0.2.0-linux-$BIN_ARCH.tar.gz"
+        tar -xzf kuadrantctl-v0.2.0-linux-$BIN_ARCH.tar.gz
+    - name: run-kuadrantctl
+      image: alpine:latest
```
Comment (jasonmadigan) [inline doc/kuadrantctl-ci-cd.md:86]: might be worth considering a kudrantctl image at some point with this stuff baked in.Could merge/shorten steps 3/4

File: doc/kuadrantctl-ci-cd.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
+# kuadrantctl - CI/CD with Tekton and Argo CD
+
+This guide demonstrates setting up a CI/CD pipeline using Tekton to deploy Kubernetes Gateway API and Kuadrant resources generated by `kuadrantctl`, from an OpenAPI specification. In this example, these resources are applied directly to the cluster where Tekton is running.
+
+Prerequisites:
+
+- (Tekton Pipelines[https://tekton.dev/]) installed on your Kubernetes or OpenShift cluster.
+- (`kubectl`[https://kubernetes.io/docs/reference/kubectl/]) configured to communicate with your cluster (i.e you have a kubectl config available with access to your cluster)
+- (Tekton CLI `tkn`[https://tekton.dev/docs/cli/]) (optional) for easier interaction with Tekton resources.
```
Comment (david-martin) [inline doc/kuadrantctl-ci-cd.md:9]: Worth adding Kuadrant installed (and gateway api crds)?

**Review by david-martin — APPROVED** [review]
Small suggestion for pre-reqs, otherwise all good.

```bash
tkn tr list -n petstore
NAME                      STARTED          DURATION   STATUS
run-kuadrantctl-taskrun   32 seconds ago   17s        Succeeded
```

```bash
tkn tr logs run-kuadrantctl-taskrun -f -n petstore

[clone] Cloning into '/workspace/source'...
[clone] Already on 'main'
[clone] Your branch is up to date with 'origin/main'.

[download-kuadrantctl]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[download-kuadrantctl]                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 21.4M  100 21.4M    0     0  12.7M      0  0:00:01  0:00:01 --:--:-- 21.2M

[run-kuadrantctl] kind: AuthPolicy
[run-kuadrantctl] apiVersion: kuadrant.io/v1beta2
[run-kuadrantctl] metadata:
[run-kuadrantctl]   name: petstore
[run-kuadrantctl]   namespace: petstore
[run-kuadrantctl]   creationTimestamp: null
[run-kuadrantctl]   labels:
[run-kuadrantctl]     deployment: petstore
[run-kuadrantctl]     owner: jbloggs
[run-kuadrantctl] spec:
[run-kuadrantctl]   targetRef:
[run-kuadrantctl]     group: gateway.networking.k8s.io
[run-kuadrantctl]     kind: HTTPRoute
[run-kuadrantctl]     name: petstore
[run-kuadrantctl]     namespace: petstore
[run-kuadrantctl]   routeSelectors:
[run-kuadrantctl]     - matches:
[run-kuadrantctl]         - path:
[run-kuadrantctl]             type: Exact
[run-kuadrantctl]             value: /api/v3/store/admin
[run-kuadrantctl]           method: GET
[run-kuadrantctl]   rules:
[run-kuadrantctl]     authentication:
[run-kuadrantctl]       storeAdmin_api_key:
[run-kuadrantctl]         credentials:
[run-kuadrantctl]           customHeader:
[run-kuadrantctl]             name: api_key
[run-kuadrantctl]         apiKey:
[run-kuadrantctl]           selector:
[run-kuadrantctl]             matchLabels:
[run-kuadrantctl]               kuadrant.io/apikeys-by: api_key
[run-kuadrantctl]         routeSelectors:
[run-kuadrantctl]           - matches:
[run-kuadrantctl]               - path:
[run-kuadrantctl]                   type: Exact
[run-kuadrantctl]                   value: /api/v3/store/admin
[run-kuadrantctl]                 method: GET
[run-kuadrantctl] status: {}
[run-kuadrantctl] kind: RateLimitPolicy
[run-kuadrantctl] apiVersion: kuadrant.io/v1beta2
[run-kuadrantctl] metadata:
[run-kuadrantctl]   name: petstore
[run-kuadrantctl]   namespace: petstore
[run-kuadrantctl]   creationTimestamp: null
[run-kuadrantctl]   labels:
[run-kuadrantctl]     deployment: petstore
[run-kuadrantctl]     owner: jbloggs
[run-kuadrantctl] spec:
[run-kuadrantctl]   targetRef:
[run-kuadrantctl]     group: gateway.networking.k8s.io
[run-kuadrantctl]     kind: HTTPRoute
[run-kuadrantctl]     name: petstore
[run-kuadrantctl]     namespace: petstore
[run-kuadrantctl]   limits:
[run-kuadrantctl]     getInventory:
[run-kuadrantctl]       routeSelectors:
[run-kuadrantctl]         - matches:
[run-kuadrantctl]             - path:
[run-kuadrantctl]                 type: Exact
[run-kuadrantctl]                 value: /api/v3/store/inventory
[run-kuadrantctl]               method: GET
[run-kuadrantctl]       rates:
[run-kuadrantctl]         - limit: 10
[run-kuadrantctl]           duration: 10
[run-kuadrantctl]           unit: second
[run-kuadrantctl]     loginUser:
[run-kuadrantctl]       routeSelectors:
[run-kuadrantctl]         - matches:
[run-kuadrantctl]             - path:
[run-kuadrantctl]                 type: Exact
[run-kuadrantctl]                 value: /api/v3/user/login
[run-kuadrantctl]               method: GET
[run-kuadrantctl]       rates:
[run-kuadrantctl]         - limit: 2
[run-kuadrantctl]           duration: 10
[run-kuadrantctl]           unit: second
[run-kuadrantctl] status: {}
[run-kuadrantctl] kind: HTTPRoute
[run-kuadrantctl] apiVersion: gateway.networking.k8s.io/v1beta1
[run-kuadrantctl] metadata:
[run-kuadrantctl]   name: petstore
[run-kuadrantctl]   namespace: petstore
[run-kuadrantctl]   creationTimestamp: null
[run-kuadrantctl]   labels:
[run-kuadrantctl]     deployment: petstore
[run-kuadrantctl]     owner: jbloggs
[run-kuadrantctl] spec:
[run-kuadrantctl]   parentRefs:
[run-kuadrantctl]     - kind: Gateway
[run-kuadrantctl]       namespace: kuadrant-multi-cluster-gateways
[run-kuadrantctl]       name: prod-web
[run-kuadrantctl]   hostnames:
[run-kuadrantctl]     - petstore.${KUADRANT_ZONE_ROOT_DOMAIN}
[run-kuadrantctl]   rules:
[run-kuadrantctl]     - matches:
[run-kuadrantctl]         - path:
[run-kuadrantctl]             type: Exact
[run-kuadrantctl]             value: /api/v3/store/admin
[run-kuadrantctl]           method: GET
[run-kuadrantctl]       backendRefs:
[run-kuadrantctl]         - name: petstore
[run-kuadrantctl]           namespace: petstore
[run-kuadrantctl]           port: 8080
[run-kuadrantctl]     - matches:
[run-kuadrantctl]         - path:
[run-kuadrantctl]             type: Exact
[run-kuadrantctl]             value: /api/v3/store/inventory
[run-kuadrantctl]           method: GET
[run-kuadrantctl]       backendRefs:
[run-kuadrantctl]         - name: petstore
[run-kuadrantctl]           namespace: petstore
[run-kuadrantctl]           port: 8080
[run-kuadrantctl]     - matches:
[run-kuadrantctl]         - path:
[run-kuadrantctl]             type: Exact
[run-kuadrantctl]             value: /api/v3/user/login
[run-kuadrantctl]           method: GET
[run-kuadrantctl]       backendRefs:
[run-kuadrantctl]         - name: petstore
[run-kuadrantctl]           namespace: petstore
[run-kuadrantctl]           port: 8080
[run-kuadrantctl] status:
[run-kuadrantctl]   parents: null

[apply-resources] authpolicy.kuadrant.io/petstore unchanged
[apply-resources] httproute.gateway.networking.k8s.io/petstore created
[apply-resources] ratelimitpolicy.kuadrant.io/petstore created
```

**Pull request merged**